MSC4222: Adding `state_after` to sync v2 (#4222)
matrix-org/matrix-spec-proposals: proposals/4222-sync-v2-state-after.md

Reviewer: Implementation requirements:
* [ ] Server
* [ ] Client

This is basically the same as how state is returned in [MSC4186 - Simplified Sliding
Sync](https://github.com/matrix-org/matrix-spec-proposals/pull/4186).

State events that appear in the timeline section **MUST NOT** update the current state. The current state **MUST** only be
updated with the contents of `state_after`.

Clients can tell if the server supports this change by whether it returns a `state` or `state_after` section in the
response.

### Examples

#### Example 1 \- Common case

Let’s take a look at the common case of a state event getting sent down an incremental sync, which is non-gappy.

<table>
<tr><th>Previously</th><th>Proposed</th></tr>
<tr>
<td>

```json
{
  "timeline": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ],
    "limited": false,
  },
  "state": {
    "events": []
  }
}
```

</td>
<td>

```json
{
  "timeline": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ],
    "limited": false,
  },
  "state_after": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ]
  }

> ```

Author: well that didn't help :angry:

> ```

Author: oh it was missing a bracket!

</td>
</tr>
</table>

Since the current state of the room will include the new state event, it's included in the `state_after` section.

> [!NOTE]
> In the proposed API the state event comes down both in the timeline section *and* the state section.

#### Example 2 - Receiving “outdated” state

Next, let’s look at what would happen if we receive a state event that does not take effect, i.e. that shouldn’t cause the client to update its state.

<table>
<tr><th>Previously</th><th>Proposed</th></tr>
<tr>
<td>

```json
{
  "timeline": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ],
    "limited": false,
  },
  "state": {
    "events": []
  }
}
```

</td>
<td>

```json
{
  "timeline": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ],
    "limited": false,
  },
  "state_after": {
    "events": []
  }
}
```

</td>
</tr>
</table>

Since the current state of the room does not include the new state event, it's excluded from the `state_after` section.

> [!IMPORTANT]
> Both responses are the same, but the client **MUST NOT** update its state with the event.

> ## Potential issues

Reviewer 2: > The format of returned state in `required_state` is a list of events. This does now allow the server to indicate if a "state reset" has happened which removed an entry from the state entirely (rather than it being replaced with another event).

Does this issue from [MSC4186: Simplified Sliding Sync](https://github.com/matrix-org/matrix-spec-proposals/pull/4186)  apply to this MSC also

Author: Mmm, that is true. I guess there is an open question as to whether we want to try and fix that here or not :thinking:

Reviewer 2: Yeah ok but is this easy or hard to fix? If it isn't easy, then will it be harder to fix when the proposal is spec? I don't mind a compromise as long as we're not hurting ourselves later

Reviewer 3: It depends how you classify easy or hard :D

It's technically very easy, as you just need to have some representation of "deleted" events: we've tossed around `content: {}` and `content: null` and a few others like having a new key in the response JSON. It's socially difficult however as it's prone to bikeshedding.

Reviewer 2: Ah, i see. `content: {}` is a bit more palatable for me since that's already how a lot of tools send state events to "remove" them from current state, and it doesn't break anyone's event schema. I'd probably pair it with a flag in unsigned or something though.

Reviewer 2: Do you intend to show if a state event in current state has been redacted? How would that work?

What about killing two birds in one stone and repeat the event with empty content. Since then it doesn't matter if a client is naive and unaware of the behavior, if they naively update their model then they'll end up with the right state.

Reviewer 2: > 1. Mixing deleted state in with the other state means that clients essentially need to handle multiple datatypes in the same list.

> 2. Having a separate field will make it easy for client implementations to miss/forget implementing the extra field during implementation. From my experience, its fairly common for people not to closely read the docs when initially writing clients, and so I think this will happen more often than you might think.

I mean if you really want to avoid both of these ideally you could use a new type for state deltas with a property for the event and a property for the change type. In Draupnir land we call them: Introduced (for unseen state key type pair), Superseded, Redacted, and I guess you'd also need Removed[^1].

Not sure if it's practical or possible for you to change that now though.

[^1]: It's actually a bit more complicated because we try to not keep completely redacted state events around as an optimization and try to tell people exactly how they have changed but eh it's not necessary to copy that https://github.com/Gnuxie/matrix-protection-suite/blob/main/src/StateTracking/StateChangeType.ts#L13-L64

Reviewer 4: > Any fix we do to communicate deleted state in SSS can easily enough be backported to v2 should it be needed. I see no reason to block the MSC on this point. On the contrary, this MSC _massively_ reduces desync issues compared to v2 without `state_after` because we are now tracking diffs against the current state and not diffs against the last timeline event. For such a small change, this MSC provides immense value.

+1; I'm strongly in favour of *not* blocking this MSC for years until someone has time to wrangle this last nit. I don't think that landing it as-is will make it materially more difficult to fix later.

Reviewer 2: Might be way down on the lower end of priorities but this isn't a nit. And it shouldn't be taking years to sort issues out like this. If it is, then that's a more urgent issue to solve because then things are entirely dysfunctional.

Reviewer 4: @Gnuxie you're right, it's not a nit and it is important; I'm sorry I spoke in a dismissive tone about it. You're also right that finding a solution shouldn't take years; I very much hope a resolution can be reached in a much shorter timeframe. 

Nevertheless, given the practicalities of where we are right now in terms of time and money, I think it would be a grave mistake to let perfect be the enemy of good by holding this MSC up while we wait for a solution to come along.

With the proposed API the common case for receiving a state update will cause the event to come down in both the
`timeline` and `state` sections, potentially increasing bandwidth usage. However, it is common for the HTTP responses to
be compressed, heavily reducing the impact of having duplicated data.

> Clients will not be able to tell when a state change happened within the timeline. This was used by some clients to
> render e.g. display names of users at the time they sent the message (rather than their current display name), though
> e.g. Element clients have moved away from this UX. This behavior can be replicated in the same way that clients dealt
> with messages received via pagination (i.e. calling `/messages`), by walking the timeline backwards and inspecting the
> `unsigned.prev_state` field. While this can lead to incorrect results, this is no worse than the previous situation.

Reviewer 5: This kind of usage of the timeline section should be not impacted at all.
We still send all state events (even if they might be wrong) in the timeline section so clients using those state event to render things in the timeline should be not impacted at all.

```suggestion
As before clients will not be able to tell when a state change happened within the timeline. This was used by some clients to
render e.g. display names of users at the time they sent the message (rather than their current display name), though
e.g. Element clients have moved away from this UX.
This MSC allows to compute the current state correctly. It does not fix the behavior for the state history in the timeline. So clients using such an approach still will face the same issues they had before this MSC.

A proper solution for this usecase would be to paginate back state using the `replaces_state` id. With this one can find all actual state event ids. This list of Ids can be used in the timeline history to filter and invalid state changes.
```

Author: I guess what I wanted to make clear was that clients will need to change how they calculate the "state at each event", and that that behaviour is really considered best-effort at best.

Reviewer 6: ```suggestion
As before, clients will not be able to reliably tell when a state change happened within the timeline. Currently, some clients walk through the timeline to
render e.g. display names of users at the time they sent the message (rather than their current display name), though
e.g. Element clients have moved away from this UX.
This MSC allows to compute the current state correctly. It does not fix the behavior for the state history in the timeline. So clients using such an approach still will face the same issues they had before this MSC.
```

# MSC4222: Adding `state_after` to sync v2

The current sync v2 API does not differentiate between state events in the timeline and updates to state, and so can
cause the client's view of the current state of the room to diverge from the actual state of the room. This is
particularly problematic for use-cases that rely on state being consistent between different clients.

This behavior stems from the fact that the clients update their view of the current state with state events that appear
in the timeline. To handle gappy syncs, the `state` section includes state events that are from *before* the start of
the timeline, and so are replaced by any matching state events in the timeline. This provides little opportunity for the
server to ensure that the clients come to the correct conclusion about the current state of the room.

In [MSC4186 - Simplified Sliding Sync](https://github.com/matrix-org/matrix-spec-proposals/pull/4186) this problem is
solved by the equivalent `required_state` section including all state changes between the previous sync and the end of

> the current sync, and clients do not update their view of state based on entries in the timeline.

Reviewer 3: The problem is that clients need to know the current state and the current model cannot express the current state. Moving to "state_after" is perhaps a bit of a misnomer as it's actually moving to the _current state_ instead. A trivial example would be this graph (all events are non-conflicted state events):
```
   A
   |
   B
  / \
 C   D

State after C = {A,B,C}
State after D = {A,B,D}
Current State = {A,B,C,D}
```

Implementations which try to think of this as "some start state" + "some linear sequence of events"  = "end state" are fundamentally calculating the wrong thing because forks in the graph (concurrent events) can occur at any point on the linear sequence of events. You might argue that this is commutative so just union the lot (i.e a timeline of [A,B,D,C] or [A,B,C,D] are valid arrival orders and hey the union of either is {A,B,C,D} but the problem is when you then introduce state resolution which is effectively a sorting and filtering algorithm:

```
   A
   |
   B
  / \
 C   C'   <-- C' will win this
  \ / \
   D   E

State after C = {A,B,C}
State after D = {A,B,C'}
State after E = {A,B,C',E}
Current State = {A,B,C',D,E}
```

Now assume you see [A,B,C',E]. The current "state before" does the right thing as does `state_after` E and the "current state" are in agreement as we haven't seen any concurrent events yet: all 3 calculations agree the end state is {A,B,C',E}. Now you get [C,D], and things get _really_ weird:
 - The client thinks the room state is `{A,B,C',E}`.
 - The current /sync impl will have a `state` block of {B} and then have `[C,D]`, meaning we incorrectly replace C' with C.
 - The true state after the last event in the timeline (D) is {A,B,C'} which is wrong as we're missing E. Thankfully, this [isn't what the implementation is doing](https://github.com/element-hq/synapse/pull/17888/files#diff-70c84622edff898486bdc02d7a579db4e52a3e0b26c14020bed0ca1fb8ea2018R1387).
 - The server impl for this MSC is calculating the current state delta, so the delta to {A,B,C',D,E} which is {D} which is correct.

When writing this up, some of this was surprising even to me, so I've backed this up with a complement test you can run which demonstrates the above fork problem in the current sync impl:

<details>
  <summary>Complement test</summary>

```go

func TestSyncDoesRightDelta(t *testing.T) {
	deployment := complement.Deploy(t, 1)
	defer deployment.Destroy(t)
	srv := federation.NewServer(t, deployment,
		federation.HandleKeyRequests(),
		federation.HandleMakeSendJoinRequests(),
		federation.HandleTransactionRequests(nil, nil),
	)
	srv.UnexpectedRequestsAreErrors = false
	cancel := srv.Listen()
	defer cancel()

	alice := deployment.Register(t, "hs1", helpers.RegistrationOpts{})
	bob := srv.UserID("bob")
	ver := alice.GetDefaultRoomVersion(t)
	serverRoom := srv.MustMakeRoom(t, ver, federation.InitialRoomEvents(ver, bob))

	// Join Alice to the new room on the federation server and make Alice a moderator
	alice.MustJoinRoom(t, serverRoom.RoomID, []string{srv.ServerName()})
	since := alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(alice.UserID, serverRoom.RoomID))

	// send the linear sequence B, C', E
	eventB := srv.MustCreateEvent(t, serverRoom, federation.Event{
		Type:     "m.room.B",
		StateKey: b.Ptr(""),
		Sender:   bob,
		Content: map[string]interface{}{
			"event": "B",
		},
	})
	// we need to fudge the origin_server_ts so we win state res, so we need a custom function
	createEventWithTimestamp := func(t *testing.T, s *federation.Server, r *federation.ServerRoom, ev federation.Event, timestamp time.Time) gomatrixserverlib.PDU {
		proto, err := serverRoom.ProtoEventCreator(serverRoom, ev)
		must.NotError(t, "ProtoEventCreator: failed to create proto event", err)
		verImpl, err := gomatrixserverlib.GetRoomVersion(r.Version)
		must.NotError(t, "GetRoomVersion: unknown version", err)
		eb := verImpl.NewEventBuilderFromProtoEvent(proto)
		signedEvent, err := eb.Build(timestamp, spec.ServerName(s.ServerName()), s.KeyID, s.Priv)
		must.NotError(t, "Build", err)
		return signedEvent
	}

	eventCPrime := createEventWithTimestamp(t, srv, serverRoom, federation.Event{
		Type:     "m.room.C",
		StateKey: b.Ptr(""),
		Sender:   bob,
		Content: map[string]interface{}{
			"event": "C'",
		},
		PrevEvents: []string{eventB.EventID()},
	}, time.Now().Add(time.Hour))

	eventE := srv.MustCreateEvent(t, serverRoom, federation.Event{
		Type:     "m.room.E",
		StateKey: b.Ptr(""),
		Sender:   bob,
		Content: map[string]interface{}{
			"event": "E",
		},
		PrevEvents: []string{eventCPrime.EventID()},
	})
	srv.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{eventB.JSON(), eventCPrime.JSON(), eventE.JSON()}, nil)
	time.Sleep(time.Second) // let the events be processed, we could use a sentinel but it's more code and this is just an example

	currentState := make(map[string]string)
	roomID := serverRoom.RoomID

	getCurrentStateForC := func() {
		stateResp := alice.Do(t, "GET", []string{"_matrix", "client", "v3", "rooms", roomID, "state"})
		body := must.ParseJSON(t, stateResp.Body)
		for _, ev := range body.Array() {
			if ev.Get("type").Str == "m.room.C" {
				t.Logf("/state says the current state of C is %v", ev.Get("content").Raw)
			}
		}
	}

	syncUpdatingStateUntilEvent := func(finalEventID string) client.SyncCheckOpt {
		return func(clientUserID string, topLevelSyncJSON gjson.Result) error {
			t.Helper()
			stateField := "rooms.join." + client.GjsonEscape(roomID) + ".state.events"
			timelineField := "rooms.join." + client.GjsonEscape(roomID) + ".timeline.events"
			if len(topLevelSyncJSON.Get(stateField).Array()) > 0 {
				t.Logf("received state field: %v", topLevelSyncJSON.Get(stateField).Raw)
			}
			timeline := topLevelSyncJSON.Get(timelineField).Array()
			for _, ev := range timeline {
				if !ev.Get("state_key").Exists() {
					continue
				}
				currentState[ev.Get("type").Str+"/"+ev.Get("state_key").Str] = ev.Get("content").Raw
				if ev.Get("event_id").Str == finalEventID {
					return nil
				}
			}
			return fmt.Errorf("not seen final event in timeline yet")
		}
	}
	since = alice.MustSyncUntil(t, client.SyncReq{Since: since}, syncUpdatingStateUntilEvent(eventE.EventID()))
	t.Logf("client calculated current state before sending C is: %+v", currentState)
	getCurrentStateForC()

	// now send C,D
	eventC := srv.MustCreateEvent(t, serverRoom, federation.Event{
		Type:     "m.room.C",
		StateKey: b.Ptr(""),
		Sender:   bob,
		Content: map[string]interface{}{
			"event": "C",
		},
		PrevEvents: []string{eventB.EventID()},
	})

	eventD := srv.MustCreateEvent(t, serverRoom, federation.Event{
		Type:     "m.room.D",
		StateKey: b.Ptr(""),
		Sender:   bob,
		Content: map[string]interface{}{
			"event": "D",
		},
		PrevEvents: []string{eventC.EventID(), eventCPrime.EventID()},
	})

	srv.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{eventC.JSON(), eventD.JSON()}, nil)
	time.Sleep(time.Second)

	// now sync until we see event D
	since = alice.MustSyncUntil(t, client.SyncReq{Since: since}, syncUpdatingStateUntilEvent(eventD.EventID()))
	t.Logf("client calculated current state after sending C is: %+v", currentState)
	getCurrentStateForC()

	t.Logf("event B: %v prevs=%v", eventB.EventID(), eventB.PrevEventIDs())
	t.Logf("event C: %v prevs=%v", eventC.EventID(), eventC.PrevEventIDs())
	t.Logf("event C': %v prevs=%v", eventCPrime.EventID(), eventCPrime.PrevEventIDs())
	t.Logf("event D: %v prevs=%v", eventD.EventID(), eventD.PrevEventIDs())
	t.Logf("event E: %v prevs=%v", eventE.EventID(), eventE.PrevEventIDs())
	must.Equal(t, currentState["m.room.C/"], `{"event":"C'"}`, "incorrect value for C")
}
```
</details>

This will produce:
```
    federation_sync_test.go:227: client calculated current state before sending C is: map[m.room.B/:{"event":"B"} m.room.C/:{"event":"C'"} m.room.E/:{"event":"E"}]
    client.go:706: [CSAPI] GET hs1/_matrix/client/v3/rooms/!0-f0MIgRBjwmRJcYz9xV:host.docker.internal:61064/state => 200 OK (3.088667ms)
    federation_sync_test.go:200: /state says the current state of C is {"event":"C'"}
    client.go:706: [CSAPI] GET hs1/_matrix/client/v3/sync => 200 OK (14.688584ms)
    federation_sync_test.go:256: client calculated current state after sending C is: map[m.room.B/:{"event":"B"} m.room.C/:{"event":"C"} m.room.D/:{"event":"D"} m.room.E/:{"event":"E"}]
    client.go:706: [CSAPI] GET hs1/_matrix/client/v3/rooms/!0-f0MIgRBjwmRJcYz9xV:host.docker.internal:61064/state => 200 OK (2.970042ms)
    federation_sync_test.go:200: /state says the current state of C is {"event":"C'"}
    federation_sync_test.go:259: event B: $TZcrwJhi0tgbUq5xtiwu9blCvhWWzetPu7h-azGdNmo prevs=[$oTThzswBLFz3b6mE8e-uDqVNgVYto6nYZsUgNOcNT88]
    federation_sync_test.go:260: event C: $KXHXQy8AgEWBTfy-7e5g5nuytM6PyY3QmtON3gGxEYo prevs=[$TZcrwJhi0tgbUq5xtiwu9blCvhWWzetPu7h-azGdNmo]
    federation_sync_test.go:261: event C': $huJSetAOxfYS3pCh28GaZc5QEgJhkmcEB2eijzSyvNw prevs=[$TZcrwJhi0tgbUq5xtiwu9blCvhWWzetPu7h-azGdNmo]
    federation_sync_test.go:262: event D: $EEzR4Ky58G4Xs4k7BvnM4ZfxWGeVOUOvTuS9vf9T9VI prevs=[$KXHXQy8AgEWBTfy-7e5g5nuytM6PyY3QmtON3gGxEYo $huJSetAOxfYS3pCh28GaZc5QEgJhkmcEB2eijzSyvNw]
    federation_sync_test.go:263: event E: $r5fBu3cXBYuJKcodxMz3QkxzOPKw3XIval1hONFOgno prevs=[$huJSetAOxfYS3pCh28GaZc5QEgJhkmcEB2eijzSyvNw]
    federation_sync_test.go:264: Equal incorrect value for C: got '{"event":"C"}' want '{"event":"C'"}'
```

> the current sync, and clients do not update their view of state based on entries in the timeline.

Reviewer 4: That's a great write-up, kegan, thanks! It would be great if it could go somewhere it can be linked from the body of the MSC (or even be added as an appendix to the MSC body?)

> the current sync, and clients do not update their view of state based on entries in the timeline.

Reviewer 6: Addressed by https://github.com/matrix-org/matrix-spec-proposals/pull/4222/commits/8d4d12317781f0f4e6eee74ab6aec33363327be0

## Proposal

> This change is gated behind the client adding a `?use_state_after=true` (the unstable name is
> `org.matrix.use_state_after`) query param.
> 
> When enabled, the Homeserver will **omit** the `state` section in the room response sections. This is replaced by
> `state_after` (the unstable field name is `org.matrix.state_after`), which will include all state changes between the

Reviewer 7: According to the "Unstable prefix" section below.

```suggestion
This change is gated behind the client adding a `?use_state_after=true` (the unstable name is
`org.matrix.msc4222.use_state_after`) query param.

When enabled, the Homeserver will **omit** the `state` section in the room response sections. This is replaced by
`state_after` (the unstable field name is `org.matrix.msc4222.use_state_after`), which will include all state changes between the
```

previous sync and the *end* of the timeline section of the current sync. This is in contrast to the old `state` section
that only included state changes between the previous sync and the *start* of the timeline section. Note that this does
mean that a new state event will (likely) appear in both the timeline and state sections of the response.

This is basically the same as how state is returned in [MSC4186 - Simplified Sliding
Sync](https://github.com/matrix-org/matrix-spec-proposals/pull/4186).

> State events that appear in the timeline section **MUST NOT** update the current state. The current state **MUST** only be

Reviewer 4: ```suggestion
Clients **MUST NOT** update their view of the current state based on events that appear in the timeline section.
The current state view **MUST** only be
```

updated with the contents of `state_after`.

Clients can tell if the server supports this change by whether it returns a `state` or `state_after` section in the

> response.

Reviewer 8: `state` isn't currently required in sync responses, but I guess `state_after` will be required in all joined rooms after this change? Should probably make it more explicit that servers MUST return `state_after` with an empty `events` array inside

Reviewer 4: ```suggestion
response. The `state_after` property is **required** by servers that support this change.
```

    } ],
    "limited": false,
  },
  "state_after": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ]
  }
}
```

</td>
</tr>
</table>

Since the current state of the room will include the new state event, it's included in the `state_after` section.

> [!NOTE]
> In the proposed API the state event comes down both in the timeline section *and* the state section.

#### Example 2 - Receiving “outdated” state

Next, let’s look at what would happen if we receive a state event that does not take effect, i.e. that shouldn’t cause the client to update its state.

<table>
<tr><th>Previously</th><th>Proposed</th></tr>
<tr>
<td>

```json
{
  "timeline": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ],
    "limited": false,
  },
  "state": {
    "events": []
  }
}
```

</td>
<td>

```json
{
  "timeline": {
    "events": [ {
      "type": "org.matrix.example",
      "state_key": ""
    } ],
    "limited": false,
  },
  "state_after": {
    "events": []
  }
}
```

</td>
</tr>
</table>

Since the current state of the room does not include the new state event, it's excluded from the `state_after` section.

> [!IMPORTANT]

> > Both responses are the same, but the client **MUST NOT** update its state with the event.

Reviewer 4: ```suggestion
> Both responses are the same, other than the change of `state` to `state_after`, but the client **MUST NOT** update its state with the event.
```

Reviewer 6: ```suggestion
> Both responses are the same, other than the change of `state` to `state_after`, but the client **MUST NOT** update its state with the event from the timeline section.
```

## Potential issues

With the proposed API the common case for receiving a state update will cause the event to come down in both the

> `timeline` and `state` sections, potentially increasing bandwidth usage. However, it is common for the HTTP responses to

Reviewer 5: @erikjohnston I agree this can get committed, we do mention explicitly that the HS has to omit the `state` section (and replace it with `state_after`).

be compressed, heavily reducing the impact of having duplicated data.

Clients will not be able to tell when a state change happened within the timeline. This was used by some clients to
render e.g. display names of users at the time they sent the message (rather than their current display name), though
e.g. Element clients have moved away from this UX. This behavior can be replicated in the same way that clients dealt
with messages received via pagination (i.e. calling `/messages`), by walking the timeline backwards and inspecting the
`unsigned.prev_state` field. While this can lead to incorrect results, this is no worse than the previous situation.

## Alternatives

There are a number of options for encoding the same information in different ways, for example the response could
include both the `state` and a `state_delta` section, where `state_delta` would be any changes that needed to be applied
to the client calculated state to correct it. However, since

> [MSC4186](https://github.com/matrix-org/matrix-spec-proposals/pull/4186) is likely to replace the sync v2 API, we may as

Reviewer 4: ```suggestion
[MSC4186](https://github.com/matrix-org/matrix-spec-proposals/pull/4186) is likely to replace the current `/sync` API, we may as
```

well use the same mechanism. This also has the benefit of showing that the proposed API shape can be successfully
implemented by clients, as the MSC is implemented and in use by clients.

Another option would be for server implementations to try and fudge the state and timeline responses to ensure that
clients came to the correct view of state. For example, if the server detects that a sync response will cause the client
to come to an incorrect view of state it could either a) "fixup" the state in the `state` section of the *next* sync
response, or b) remove or add old state events to the timeline section. While both these approaches are viable, they're
both suboptimal to just telling the client the correct information in the first place. Since clients will need to be
updated to handle the new behavior for future sync APIs anyway, there is little benefit from not updating clients now.

We could also do nothing, and instead wait for [MSC4186](https://github.com/matrix-org/matrix-spec-proposals/pull/4186)
(or equivalent) to land and for clients to update to it.

## Security considerations

> There are no security concerns with this proposal, as it simply encodes the same information sent do clients in a

Reviewer 7: ```suggestion
There are no security concerns with this proposal, as it simply encodes the same information sent to clients in a
```

different way

## Unstable prefix

| Name | Stable prefix | Unstable prefix |
| - | - | - |
| Query param | `use_state_after` | `org.matrix.use_state_after` |

> | Room response field | `state_after` | `org.matrix.state_after` |

Reviewer 4: unstable prefixes normally have the msc number in?
